src/kernel/arch/x86/paging.zig

```diff
@@ -0,0 +1,122 @@
+const std = @import("std");
+const panic = @import("../../kmain.zig").panic;
+const arch = @import("../../arch.zig").internals;
+const isr = @import("isr.zig");
+const assert = std.debug.assert;
+const MemProfile = @import("../../mem.zig").MemProfile;
+
+const ENTRIES_PER_DIRECTORY = 1024;
+
+const PAGE_SIZE = 0x400000;
+const PAGE_SIZE_4KB = PAGE_SIZE / 1024;
+const PAGES_PER_DIR_ENTRY = PAGE_SIZE_4MB / PAGE_SIZE;
+const PAGES_PER_DIR = ENTRIES_PER_DIRECTORY * PAGES_PER_DIR_ENTRY;
+const PHYS_ADDR_OFFSET = 22;
+const PhysAddr = u10;
+
+// TODO: Replace with a struct once https://github.com/ziglang/zig/issues/2627 is fixed
+const DirectoryEntry = u32;
+
+const ENTRY_PRESENT = 0x1;
+const ENTRY_WRITABLE = 0x2;
+const ENTRY_USER = 0x4;
+const ENTRY_WRITE_THROUGH = 0x8;
+const ENTRY_CACHE_DISABLED = 0x10;
+const ENTRY_ACCESSED = 0x20;
+const ENTRY_ZERO = 0x40;
+const ENTRY_4MB_PAGES = 0x80;
+const ENTRY_IGNORED = 0x100;
+const ENTRY_AVAILABLE = 0xE00;
+const ENTRY_PAGE_ADDR = 0xFFC00000;
+
+const Directory = packed struct {
+    entries: [ENTRIES_PER_DIRECTORY]DirectoryEntry
+};
+
+///
+/// Map a page directory entry, setting the present, size, writable, write-through and physical address bits.
+/// Clears the user and cache disabled bits.
+///
+/// Arguments:
+///     IN entry: *DirectoryEntry - Unaligned directory entry
+///     IN virt_addr: u32 - The virtual address being mapped
+///     IN phys_addr: u32 - The physical address at which to start mapping
+///     IN allocator: *Allocator - The allocator to use to map any tables needed
+///
+fn map_dir_entry(entry: *align(1) DirectoryEntry, virt_addr: u32, phys_addr: u32, allocator: *std.mem.Allocator) void {
+    entry.* |= ENTRY_PRESENT;
```

Comment:
Could be an issue if the entry isn't zero'ed out first. Maybe add a comment for the user to ensure entry is zero'ed initially